sdk/typespec/typespec_client_core/src/http/response.rs

```diff
@@ -45,193 +99,125 @@ impl RawResponse {
         (self.status, self.headers, self.body)
     }
 
-    /// Consume the HTTP response and return the HTTP body bytes.
+    /// Fetches the entire body and returns it as raw bytes.
+    ///
+    /// This method will force the entire body to be downloaded from the server and consume the response.
+    /// If you want to parse the body into a type, use [`read_body`](Response::deserialize_body) instead.
     pub fn into_body(self) -> ResponseBody {
         self.body
     }
 
-    /// Get the response body as the specified type from JSON.
-    pub async fn json<T>(self) -> crate::Result<T>
-    where
-        T: DeserializeOwned,
-    {
-        self.into_body().json().await
-    }
-
-    /// Get the response body as the specified type from XML.
-    #[cfg(feature = "xml")]
-    pub async fn xml<T>(self) -> crate::Result<T>
-    where
-        T: DeserializeOwned,
-    {
-        self.into_body().xml().await
+    /// Fetches the entire body and tries to convert it into type `U`.
+    ///
+    /// This method is intended for use in rare cases where the body of a service response should be parsed into a user-provided type.
+    ///
+    /// # Example
+    /// ```rust
+    /// # pub struct GetSecretResponse { }
+    /// use typespec_client_core::http::Response;
+    /// use serde::Deserialize;
+    /// use bytes::Bytes;
+    ///
+    /// #[derive(Deserialize)]
+    /// struct MySecretResponse {
+    ///    value: String,
+    /// }
+    /// typespec_client_core::json_model!(MySecretResponse); // Mark the type as deserializable from JSON.
+    ///
+    /// async fn parse_response(response: Response<GetSecretResponse>) {
+    ///   // Calling `map` will parse the body into `MySecretResponse` instead of `GetSecretResponse`.
+    ///   let my_struct: MySecretResponse = response.deserialize_body_into().await.unwrap();
+    ///   println!("value: {}", my_struct.value);
+    /// }
+    /// ```
+    pub async fn deserialize_body_into<U: Model>(self) -> crate::Result<U> {
+        U::from_response_body(self.body).await
     }
 }
 
-impl fmt::Debug for RawResponse {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Response")
-            .field("status", &self.status)
-            .field("headers", &self.headers)
-            .field("body", &"(body)")
-            .finish()
-    }
-}
-
-impl<T> From<RawResponse> for Response<T> {
-    fn from(response: RawResponse) -> Self {
-        Self {
-            response,
+impl Response<()> {
+    /// Changes the type of the response body.
+    ///
+    /// Used to set the "type" of an untyped `Response<()>`, transforming it into a `Response<T>`.
+    pub fn map_body<T>(self) -> Response<T> {
```

Comment:
I don't want to block the PR on this - but there needs to be another commit for the formatting issues, so I leave it up to you whether you want to change this now...if we can agree. I think "map" feels right, but "body" doesn't given it maps a Response<()> into a `Response<T>`. What about just implementing `Into` in this case?

```rust
use std::marker::PhantomData;

#[derive(serde::Deserialize)]
struct Response<'a, T: serde::Deserialize<'a>> {
    bytes: Vec<u8>,
    phantom: PhantomData<&'a T>,
}

impl<'a, T> Into<Result<Response<'a, T>, ()>> for Response<'a, ()>
where
    T: serde::Deserialize<'a>,
{
    fn into(self) -> Result<Response<'a, T>, ()> {
        serde_json::from_slice(&self.bytes).map_err(|_| ())
    }
}
```

Reply:
Unfortunately no, we can't use `Into` or `From` here. Both of `impl<T> From<Response<()> for Response<T>` **and** `impl<T> Into<Response<T>> for Response<()>` conflict with existing impls (`impl<T> From<T> for T`).

However, let me see if I can make this `pub(crate)`. I think I can now, and doing that means the weirdness is limited to `typespect_client_core`. A service client is going to call `Pipeline::send<T>` which already returns a `Response<T>` (because it calls `map_body` within it). For the rare service client that needs to take the "raw" `Response<()>` returned by a policy or HTTP client and deserialize it into a custom model (i.e. Azure Identity), they can use `deserialize_body_into`, which exists on `Response<()>`.

Reply:
This works for now, though the `set_` prefix seems odd for something that isn't a setter. I'll let you decide whether to act now, later, or not at all (given we'll probably rename a bunch of this later anyway, depending on how #1766 goes) before merging.

Reply:
That's true. How about `with_`?

Reply:
Sounds better, sure.